app/backend/approaches/chatreadretrieveread.py

```diff
@@ -90,94 +92,58 @@ def run(self, history: Sequence[dict[str, str]], overrides: dict[str, Any]) -> A
 
         msg_to_display = '\n\n'.join([str(message) for message in messages])
 
         return {"data_points": results, "answer": chat_content, "thoughts": f"Searched for:<br>{q}<br><br>Conversations:<br>" + msg_to_display.replace('\n', '<br>')}
     
-    def get_chat_history_as_text(self, history: Sequence[dict[str, str]], include_last_turn: bool=True, approx_max_tokens: int=1000) -> str:
+    def get_chat_history_as_text(self, history: Sequence[dict[str, str]], include_last_turn: bool=True, max_tokens: int = 4000) -> str:
         history_text = ""
         for h in reversed(history if include_last_turn else history[:-1]):
             history_text = """<|im_start|>user""" + "\n" + h["user"] + "\n" + """<|im_end|>""" + "\n" + """<|im_start|>assistant""" + "\n" + (h.get("bot", "") + """<|im_end|>""" if h.get("bot") else "") + "\n" + history_text
-            if len(history_text) > approx_max_tokens*4:
+            if len(history_text) > max_tokens:
                 break    
         return history_text
-    
-    def get_messages_from_history(self, prompt_override, follow_up_questions_prompt, history: Sequence[dict[str, str]], sources: str, approx_max_tokens: int = 1000) -> []:
-        '''
-        Generate messages needed for chat Completion api
-        '''
-        messages = []
-        token_count = 0
-        if prompt_override is None:
-            system_message = self.system_message_chat_conversation.format(injected_prompt="", follow_up_questions_prompt=follow_up_questions_prompt)
-        elif prompt_override.startswith(">>>"):
-            system_message = self.system_message_chat_conversation.format(injected_prompt=prompt_override[3:] + "\n", follow_up_questions_prompt=follow_up_questions_prompt)
-        else:
-            system_message = prompt_override.format(follow_up_questions_prompt=follow_up_questions_prompt)
 
-        messages.append({"role":self.SYSTEM, "content": system_message})
-        token_count += self.num_tokens_from_messages(messages[-1], self.chatgpt_model)
-        
-        # latest conversation
-        user_content = history[-1]["user"] + " \nSources:" + sources
-        messages.append({"role": self.USER, "content": user_content})
-        token_count += token_count + self.num_tokens_from_messages(messages[-1], self.chatgpt_model)
-
-        '''
-        Enqueue in reverse order
-        if limit exceeds truncate old messages 
-        leaving system message behind
-        Keep track of token count for each conversation
-        If token count exceeds limit, break
-        '''
+    def get_messages_from_history(self, system_prompt: str, modelid: str, history: Sequence[dict[str, str]], user_conv: str, few_shots = []) -> []:
+        message_builder = MessageBuilder(system_prompt, modelid)
+
+        #add shots
+        for shot in few_shots:
+            message_builder.append_message(shot.get('role'), shot.get('content'))
+
```

Comment:
nit: it'd be good to pass the code through some python styling/linting tool to standardize whitespace, blank lines, etc. In typescript we use prettier/eslint.